Content.Server/PowerSink/PowerSinkSystem.cs

```diff
@@ -32,24 +52,82 @@ private void OnExamine(EntityUid uid, PowerSinkComponent component, ExaminedEven
+
+                var currentBatteryThreshold = battery.CurrentCharge / battery.MaxCharge;
+
+                // Check for warning message threshold
+                if (!component.SentImminentExplosionWarningMessage &&
+                    currentBatteryThreshold >= WarningMessageThreshold)
+                {
+                    NotifyStationOfImminentExplosion(entity, component);
+                }
+
+                // Check for warning sound threshold
+                foreach (var testThreshold in WarningSoundThresholds)
+                {
+                    if (currentBatteryThreshold >= testThreshold &&
+                        testThreshold > component.HighestWarningSoundThreshold)
+                    {
+                        component.HighestWarningSoundThreshold = currentBatteryThreshold; // Don't re-play in future until next threshold hit
+                        _audio.PlayPvs(component.ElectricSound, entity,
+                            new Robust.Shared.Audio.AudioParams(1f, 1f, "Master", SharedAudioSystem.DefaultSoundRange * 5, 1, 25, false, 0f) // audible from farther than usual
+                        ); // Play SFX
+                        break;
+                    }
+                }
+
+                // Check for explosion
                 if (battery.CurrentCharge < battery.MaxCharge) continue;
 
-                toRemove.Add((comp, battery));
+                if (component.ExplosionTime == null)
+                {
+                    // Set explosion sequence to start soon
+                    component.ExplosionTime = _gameTiming.CurTime.Add(ExplosionDelayTime);
+
+                    // Wind-up SFX
+                    _audio.PlayPvs(component.ChargeFireSound, entity); // Play SFX
+                } else if (_gameTiming.CurTime >= component.ExplosionTime) {
+                    // Explode!
+                    toRemove.Add((entity, component, battery));
+                }
             }
 
-            foreach (var (comp, battery) in toRemove)
+            foreach (var (entity, component, battery) in toRemove)
             {
-                _explosionSystem.QueueExplosion(comp.Owner, "Default", 5 * (battery.MaxCharge / 2500000), 0.5f, 10, canCreateVacuum: false);
-                EntityManager.RemoveComponent(comp.Owner, comp);
+                _explosionSystem.QueueExplosion(entity, "PowerSink", 2000f, 4f, 20f, canCreateVacuum: true);
+                EntityManager.RemoveComponent(entity, component);
             }
         }
+
+        private void NotifyStationOfImminentExplosion(EntityUid uid, PowerSinkComponent powerSinkComponent)
+        {
+            if (powerSinkComponent.SentImminentExplosionWarningMessage)
+                return;
+            powerSinkComponent.SentImminentExplosionWarningMessage = true;
+
+            var stationSystem = _entitySystemManager.GetEntitySystem<StationSystem>();
+            var chatSystem = _entitySystemManager.GetEntitySystem<ChatSystem>();
```

Comment:
Just store dependencies on powersinksystem ([Dependency] attribute).

Reply:
I can do this, but just FYI since this code block is only called quite rarely, performance difference of this seems like it should be negligible.  🤷